fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/auto/DruidNavigator.java

```diff
@@ -0,0 +1,190 @@
+// Copyright 2016 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.wiki.webservice.data.config.auto;
+
+import com.yahoo.bard.webservice.data.time.DefaultTimeGrain;
+import com.yahoo.bard.webservice.data.time.GranularityDictionary;
+import com.yahoo.bard.webservice.data.time.StandardGranularityParser;
+import com.yahoo.bard.webservice.data.time.TimeGrain;
+import com.yahoo.bard.webservice.druid.client.DruidWebService;
+import com.yahoo.bard.webservice.druid.client.SuccessCallback;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * Searches for all datasources and their dimensions, metrics, and timegrains from druid.
+ */
+public class DruidNavigator implements Supplier<List<? extends DataSourceConfiguration>> {
+    private static final Logger LOG = LoggerFactory.getLogger(DruidNavigator.class);
+    private static final String DATASOURCES = "/datasources/";
+    private final DruidWebService druidWebService;
+    private final List<TableConfig> tableConfigurations;
+
+    /**
+     * Constructs a DruidNavigator to load datasources from druid.
+     *
+     * @param druidWebService The DruidWebService to be used when talking to druid.
+     */
+    public DruidNavigator(DruidWebService druidWebService) {
+        this.druidWebService = druidWebService;
+        tableConfigurations = new ArrayList<>();
+    }
+
+    /**
+     * Queries druid for all datasources and loads each of their configurations.
+     *
+     * @return a list of {@link DataSourceConfiguration} containing all known druid datasources configs.
+     */
+    @Override
+    public List<? extends DataSourceConfiguration> get() {
+        if (tableConfigurations.isEmpty()) {
+            loadAllDatasources();
+            LOG.info("Loading all datasources");
+            try {
+                //TODO Fix this so it waits for druid's responses
+                // instead of waiting 1 second for everything to be configured
+                // in order to wait for druid's response, DruidWebService will have to return a future
+                Thread.sleep(1000);
```

Comment:
Good call! I always forget about `CountDownLatch`! It's definitely the right choice here.

That said, the other option (which I think might be the route to take) would be to make the async client return a `Future` so that consumers (like this one) can switch to a blocking behavior if they wish.